src/main/java/uk/gov/pay/webhooks/validations/CallbackUrlService.java

```diff
@@ -0,0 +1,60 @@
+package uk.gov.pay.webhooks.validations;
+
+import com.google.common.net.InternetDomainName;
+import uk.gov.pay.webhooks.app.WebhooksConfig;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Optional;
+import java.util.Set;
+
+import static java.util.stream.Collectors.toUnmodifiableSet;
+
+public class CallbackUrlService {
+    private final Set<InternetDomainName> allowedDomains;
+
+    public CallbackUrlService(WebhooksConfig webhooksConfig) {
+        this.allowedDomains = webhooksConfig.getLiveDataAllowHosts().stream().map(InternetDomainName::from).collect(toUnmodifiableSet());
+    }
+
+    public void validateCallbackUrl(String callbackUrl, Boolean contextIsLive) throws MalformedURLException {
+        var url = validateAndGetUrlIsWellFormed(callbackUrl);
+        if (Boolean.TRUE.equals(contextIsLive)) {
+            validateUrlIsInLiveDomains(url);
+        }
+    }
+
+    private URL validateAndGetUrlIsWellFormed(String callbackUrl) throws MalformedURLException {
+        URL url;
+        try {
+            url = new URL(callbackUrl);
+        } catch (MalformedURLException e) {
+            throw new CallbackUrlMalformedException("Callback URL is not a valid URL");
+        }
+
+        if (Optional.ofNullable(url.getHost()).map(String::isEmpty).orElse(true)) {
+            throw new CallbackUrlMalformedException("Callback URL must contain a host");
+        }
+
+        if (!"https".equals(url.getProtocol())) {
+            throw new CallbackUrlProtocolNotSupported("Callback URL must use HTTPS protocol");
+        }
+
+        if (!InternetDomainName.isValid(url.getHost())) {
+            throw new CallbackUrlMalformedException("Callback URL is not a valid domain name");
```

Comment:
Really nitty but I think something like “Callback URL host must be a domain name” might be best since if we hit this it’s most likely that we have an IP address (an invalid domain name should not get this far).